dask_grblas/expr.py

```diff
@@ -392,25 +407,107 @@ def _extractor_new(x, dtype, mask, mask_type):
+        resolved_indices = _resolve_indices(value, indices)
+        if mask_level == 2:
+            # mask_level == 2 means that during blockwise(), dask extracted
+            # part  of the  InnerVector chunk at Extractor level 1
+            # to match grblas.Mask chunk size at Extractor level 2
+            # Now we expand the mask to the size of the InnerVector
+            # so that we can update the grblas.Vector chunk directly
+            # and not a copy of it
+            mask_indices = np.arange(value.size)[indices[-mask_level + 1][0]]
+            mask = _expand_mask_to_fit(value, mask_indices, mask, mask_type)
+
+        value(mask=mask, accum=accum)[resolved_indices] << obj.value
+    else:
+        raise NotImplementedError(f"indices: {indices}")
+    return wrap_inner(gb.Vector.new(dtype=dtype, size=obj.value.size))
+
+
+def _get_expanded_mask_indices(mask, mask_range):
+    def complement_if(mask_indices, mask):
+        if mask.complement:
+            return list(set(range(mask.mask.size)) - set(mask_indices))
+        return mask_indices
+
+    structure, values = mask.mask.to_values()
+    if mask.value:
+        mask_indices = [s for s, v in zip(structure, values) if v]
+    elif mask.structure:
+        mask_indices = structure
+    else:
+        TypeError("Mask must be a ValueMask or StructuralMask, " "complemented or not.")
+    mask_indices = complement_if(mask_indices, mask)
+    return list(set(mask_range).intersection(set(mask_indices)))
+
+
 class Extractor:
-    def __init__(self, inner, index=None):
+    """
+    This stores the extraction history of an InnerVector.
+    """
+
+    def __init__(self, inner, mask=None, mask_type=None, replace=None, index=None):
         self.inner = inner
+        self.mask = mask
         self.index = index
         self.dtype = inner.dtype
         self.ndim = inner.ndim
+        if type(inner) is Extractor:
+            self.level = inner.level + 1
+        else:
+            self.level = 0
+        if replace and (type(mask) is not np.ndarray):
+            # This branch is necessary when replace is True:
+            # Here, we remove all chunk elements in the mask
+            # range that lie in the complement of the mask
+            if self.level == 1:
+                chunk = inner.inner.value
+            elif self.level == 2:
+                chunk = inner.inner.inner.value
+
+            all_indices = np.arange(chunk.size)
+            mask_range = slice(None) if inner.index is None else inner.index[0]
+            mask_range = all_indices[mask_range]
+            mask = mask_type(mask.value)
+            mask = _expand_mask_to_fit(chunk, mask_range, mask, mask_type)
+            mask_indices = _get_expanded_mask_indices(mask, mask_range)
+            indices_to_remove = set(mask_range) - set(mask_indices)
+
+            # this loop needs to be made 'race-free'!
+            for chunk_index in indices_to_remove:
+                del chunk[chunk_index]
```

Comment:
So here I am a bit concerned about the possible occurrence of race-conditions when the chosen dask-scheduler is multi-threaded (the default on my laptop).  

So far, I haven't encountered any problems on my laptop from this piece of code though.